src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine;
+
+import com.google.cloud.tools.appengine.api.AppEngineException;
+import java.io.IOException;
+import java.io.InputStream;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import org.w3c.dom.DOMException;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+/**
+ * Utilities to obtain information from appengine-web.xml.
+ */
+public class AppEngineDescriptor {
+
+  private static final String WEB_XML_NS_URI = "http://appengine.google.com/ns/1.0";
+  private Document document;
+  
+  private AppEngineDescriptor() {
+    // force use of parse method
+  }
+
+  /**
+   * Parses an appengine-web.xml file.
+   * 
+   * @param in the contents of appengine-web.xml
+   * @return a fully parsed object that can be queried 
+   * @throws IOException if parsing fails for any reason
+   */
+  public static AppEngineDescriptor parse(InputStream in) throws IOException {
+    try {
+      AppEngineDescriptor instance = new AppEngineDescriptor();
+      DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+      documentBuilderFactory.setNamespaceAware(true);
+      instance.document = documentBuilderFactory.newDocumentBuilder().parse(in);
+      return instance;
+    } catch (SAXException | ParserConfigurationException exception) {
+      throw new IOException("Cannot parse appengine-web.xml", exception);
+    }
+  }
+
+  /**
+   * @return project ID from the &lt;application&gt; element of the appengine-web.xml or null
+   *         if it is missing
+   * @throws AppEngineException if parsing the value fails
+   */
+  public String getProjectId()  {
+    return getTopLevelValue(document, "appengine-web-app", "application");
+  }
+
+  /**
+   * @return project version from the &lt;version&gt; element of the appengine-web.xml or
+   *         null if it is missing
+   * @throws AppEngineException if parsing the value fails
+   */
+  public String getProjectVersion() {
+    return getTopLevelValue(document, "appengine-web-app", "version");
+  }
+
+  /**
+   * @return service ID from the &lt;service&gt; element of the appengine-web.xml, or
+   *         null if it is missing. Will also look at module ID.
```

Comment:
Like the other methods that use `getTopLevelValue()`, add `@throws AppEngineException if parsing the value fails` to the javadoc?